[김병희] 선택 - RoleHierarchy 리뷰 요청 드립니다.

beng9re opened the pull request:
안녕하세요 성현님 

지난 PR https://github.com/next-step/spring-security-authorization/pull/14 피드백 적용 및 계층별 인가를 RoleHierarchyIm 를 구현하였습니다.
기본적으로 HasAuthorityAuthorizationManager 에  적용해두었습니다.

리뷰 부탁 드립니다 :)

beng9re committed:
refactor: SecuredAspect 주석 처리

SecuredMethodInterceptor 가 이미 등록되어있기 때문에 제거

beng9re committed:
refactor: 피드백 반영 CheckAuthenticationFilter 추상화에 의존

beng9re committed:
refactor: 피드백 반영 ForbiddenException 예외 처리

beng9re committed:
refactor: 피드백 반영 리다이랙션, 포워드인 액션에 대응되도록 GenericFilterBean 전환

beng9re committed:
docs(추가): 요구사항 추가

beng9re committed:
feat(추가): NullRoleHierarchy 모든 권한을 허용하는 NullRoleHierarchy 추가

beng9re committed:
feat(추가): 계층별 권한을 설정하는 RoleHierarchy 추가

boorownie commented on src/main/java/nextstep/app/SecurityConfig.java:
```diff
@@ -50,9 +50,18 @@ public SecuredMethodInterceptor securedMethodInterceptor() {
     @Bean
-    public SecuredAspect securedAspect() {
-        return new SecuredAspect();
+    public RoleHierarchy roleHierarchy() {
```
요구사항 설정 👍

boorownie commented on src/main/java/nextstep/app/SecurityConfig.java:
```diff
@@ -62,6 +71,7 @@ public RequestMatcherDelegatingAuthorizationManager requestAuthorizationManager(
         mappings.add(new RequestMatcherEntry<>(new MvcRequestMatcher(HttpMethod.GET, "/members"), new HasAuthorityAuthorizationManager("ADMIN")));
         mappings.add(new RequestMatcherEntry<>(new MvcRequestMatcher(HttpMethod.GET, "/search"), new PermitAllAuthorizationManager()));
         mappings.add(new RequestMatcherEntry<>(new MvcRequestMatcher(HttpMethod.POST, "/private"), new DenyAllAuthorizationManager()));
+        mappings.add(new RequestMatcherEntry<>(new MvcRequestMatcher(HttpMethod.GET, "/hierarchy"), new HasAuthorityAuthorizationManager("USER").withRoleHierarchy(roleHierarchy())));
```
권한 계층을 조회하도록 하셨군요!
설정이 잘 되었는지 확인하기 위해 좋은 시도입니다 :)

단, roleHierarchy()을 매번 설정할 때 마다 주입해주는 형태로 구현해주셨는데
한번만 설정하는 방법은 없을 지 고민해보셔도 좋을 것 같아요!

beng9re replied:
고민 끝에 빌더 클래스를 추출하여 개선해보았습니다!

[72fe548](https://github.com/next-step/spring-security-authorization/pull/23/commits/72fe5487d07f732b75950f6ce55af19c9dabd72b) 해당 커밋에 반영하였습니다.

boorownie commented on src/main/java/nextstep/security/authorization/CheckAuthenticationFilter.java:
```diff
@@ -2,33 +2,45 @@
+
+    private static void authorizationCheck(AuthorizationDecision authorizationDecision) {
+        if (authorizationDecision.isDeny()) {
+            throw new ForbiddenException();
```
만약 인증이 되지않은 경우 401응답이 필요할 수 있는데 현재는 모든 에로 상황에서 403이 응답이 되는 것 같네요!

beng9re replied:
앗.. 그럴수도 있겠네요!

인가실패인 경우 유동적으로 처리할 수 있도록 인가 실패 핸들러 인터페이스를 정의하여 처리해보았습니다!

[fc99a7a](https://github.com/next-step/spring-security-authorization/pull/23/commits/fc99a7a726f73bf39d3a30e7f02f4d2ecab0ec4e) 해당 커밋에 반영해두었습니다

boorownie commented on src/main/java/nextstep/security/authorization/NullRoleHierarchy.java:
```diff
@@ -0,0 +1,11 @@
+
+import java.util.Collection;
+
+public class NullRoleHierarchy implements RoleHierarchy {
```
널 객체 패턴 👍

boorownie commented on src/main/java/nextstep/security/authorization/RoleHierarchyImpl.java:
```diff
@@ -0,0 +1,57 @@
+
+import java.util.*;
+
+public class RoleHierarchyImpl implements RoleHierarchy {
```
RoleHierarchy 구현 잘 해주셨습니다 👍

boorownie commented on src/main/java/nextstep/security/authorization/HasAuthorityAuthorizationManager.java:
```diff
@@ -2,8 +2,11 @@
 
+import java.util.Collection;
+
 public class HasAuthorityAuthorizationManager implements AuthorizationManager {
```
시큐리티 코드를 보면 AuthorityAuthorizationManager 와 AuthoritiesAuthorizationManager가 있는데요
이 구조를 분석해보시고 HasAuthorityAuthorizationManager 에 대입해보셔도 좋을 것 같아요.

beng9re replied:
코드 구조를 분석해보니  AuthoritiesAuthorizationManager는 여러건의 인가를 체크할 수 있는 역할,
AuthorityAuthorizationManager 는  인가정보를 설정하는 역할로 분리한것 같아요. 

해당 부분을 관점으로 리팩토링 해보았습니다

[8ab4798](https://github.com/next-step/spring-security-authorization/pull/23/commits/8ab47981b3de163a355d9c4c50ef54f1cacf497c)해당 커밋에 반영해보았습니다

boorownie commented on src/test/java/nextstep/security/authorization/RoleHierarchyImplTest.java:
```diff
@@ -0,0 +1,38 @@
+class RoleHierarchyImplTest {
+
+    @DisplayName("계층별로 허용되는 권한을 반환한다")
+    @ParameterizedTest
+    @MethodSource("계층별로_허용되는_권한")
```
Role Hierarchy 테스트 👍

boorownie reviewed (changes requested):
선택 미션까지 잘 구현해주셨습니다.
너무 잘 구현해주셔서 특별히 코멘트 드릴 부분은 없었는데요,
추가로 고민해보시면 좋을 것 같은 부분에 코멘트 남겼으니
한번 고민해보시면 좋을 것 같아요!
꼭 코드 수정을 할 필요는 없지만 보완할 부분이 있다면 보완해보셔도 좋습니다.

beng9re committed:
refactor: 인가가 안됨에 대한 핸들링을 유동적으로 해줄수 있도록 개선

beng9re committed:
refactor: 공통의 roleHierarchy 를 가질수 있도록 RequestMatcherAuthorizationManagerBuilder 클래스로 개선

beng9re committed:
refactor: HasAuthorityAuthorizationManager 리팩토링

역할 분리

AuthoritiesAuthorizationManager : 다건에 인가 정보를 체크 하는 역할
AuthorityAuthorizationManager : 인가정보를 설정하는 역할

boorownie commented on src/main/java/nextstep/app/SecurityConfig.java:
```diff
@@ -64,16 +62,17 @@ public RoleHierarchy roleHierarchy() {
-        mappings.add(new RequestMatcherEntry<>(new MvcRequestMatcher(HttpMethod.GET, "/hierarchy"), new HasAuthorityAuthorizationManager("USER").withRoleHierarchy(roleHierarchy())));
-        mappings.add(new RequestMatcherEntry<>(new AnyRequestMatcher(), new PermitAllAuthorizationManager()));
-        return new RequestMatcherDelegatingAuthorizationManager(mappings);
+        return RequestMatcherAuthorizationManagerBuilder.withRoleHierarchy(roleHierarchy())
+                .authenticated(new MvcRequestMatcher(HttpMethod.GET, "/members/me"))
+                .permitAll(new MvcRequestMatcher(HttpMethod.GET, "/search"))
+                .hasAuthority("ADMIN", new MvcRequestMatcher(HttpMethod.GET, "/members"))
+                .hasAuthority("USER", new MvcRequestMatcher(HttpMethod.POST, "/hierarchy"))
+                .denyAll(new MvcRequestMatcher(HttpMethod.POST, "/private"))
+                .permitAll(new AnyRequestMatcher())
+        .build();
```
빌더 패턴을 적용하셨군요! 재미있는 시도입니다 👍

boorownie commented on src/main/java/nextstep/security/authorization/AuthoritiesAuthorizationManager.java:
```diff
@@ -0,0 +1,44 @@
+import java.util.Collection;
+import java.util.Set;
+
+public class AuthoritiesAuthorizationManager implements AuthorizationManager<Collection<String>> {
```
AuthoritiesAuthorizationManager 분리를 통한 책임과 역할 분리 👍

boorownie commented on src/main/java/nextstep/security/authorization/CheckAuthenticationFilter.java:
```diff
@@ -2,33 +2,43 @@
-    public CheckAuthenticationFilter(RequestMatcherDelegatingAuthorizationManager requestMatcherDelegatingAuthorizationManager) {
-        this.requestMatcherDelegatingAuthorizationManager = requestMatcherDelegatingAuthorizationManager;
+    public CheckAuthenticationFilter(AuthorizationManager<HttpServletRequest> authorizationManager) {
+        this(authorizationManager, response -> response.setStatus(HttpServletResponse.SC_FORBIDDEN));
```
구현해주신 코드를 보니 ExceptionTranslationFilter에서 처리하는 것 처럼 AuthorizationDeniedHandler를 통해 에러를 처리하도록 작업해주셨군요 👍 

사실 코멘트 드리려 했던 부분은 401이 발생할 수도 있어서  403 외 다른 응답 코드가 응답될 수 있도록 처리해보는 것을 생각해보시길 의미하는 코멘트엿습니다 :)
이렇게 할 경우 AuthorizationDeniedHandler 외 다른 handler가 필요할 수 있고, 특히 Authentication 관련 예외를 처리하는 handler는 다른 인증 필터에서도 사용될 수 있겠네요!  

요구사항에는 없던 내용이지만 참고해보시길 바라는 마음에서 남긴 코멘트였어요 :)

boorownie reviewed (approved):
선택 미션까지 잘 구현해주셨습니다!
코멘트 드린 부분을 기대 이상으로 잘 구현해주셨네요 💯

인가 미션은 여기서 마무리하고 oauth2 미션으로 넘어가도 좋겠습니다!
이만 머지하겠습니다~
